### ISSUE TITLE
[FEA] Use `cudf::contains` in `left_semi_anti_join`

### ISSUE THREAD

**Issue by ttnghia** [rapidsai/cudf#11037]
The semi- and anti- joins operations are operations that rely on building a gather map by checking if rows in one table exist in the other table. Such existence checking can be done simply by calling to `cudf::contains`.

We should modify the implementation of `left_semi_anti_join` (and other joins, if possible) to use `cudf::contains`.

**Comment by jrhemstad** [comment]
I haven't looked closely, but my suspicion is the `semi_join` implementation is already closer to the desired implementation I described here: https://github.com/rapidsai/cudf/pull/10656#discussion_r888484934

So I'd be more inclined to use that as the basis upon which `contains` is implemented, but I don't have strong feelings here.

**Comment by jrhemstad** [comment]
> So basically we can't reuse the code in `*-join` implementation but have to start from scratch with the solution as you just suggested

I don't think it would need to start from scratch. I think it would just be a matter of updating the equality/hashing operators used.

@vyasr wrote the code that is there today and should be able to help.

**Comment by ttnghia** [comment]
Updating the equality/hashing operators to use the solution you suggested (https://github.com/rapidsai/cudf/pull/10656#discussion_r888484934) is almost starting from scratch :smiley:. In summary, to have *-joins and `cudf::contains` share their implementation, we need:
 * Support nested types in `cudf::contains` (this is what `cudf::contains` wants), and
 * Add `null_equality` parameter in `cudf::contains` (this is what *-joins need)

**Comment by jrhemstad** [comment]
> Updating the equality/hashing operators to use the solution you suggested ([#10656 (comment)](https://github.com/rapidsai/cudf/pull/10656#discussion_r888484934)) is almost starting from scratch

Is it? This looks awfully close to what I already described: https://github.com/rapidsai/cudf/blob/fe9a4f88713f7ee1692738ce0deb5935da362134/cpp/src/join/semi_join.cu#L92-L152

I would like the join implementation to be the source of truth. 

The implementation I described is very similar to how `inner_join`/`left_join`  are also implemented. It would be incongruous to have the implementation for `semi_join` elsewhere when in it is so similar to the implementations for the other joins.

**Comment by ttnghia** [comment]
> Is it? This looks awfully close to what I already described:
> https://github.com/rapidsai/cudf/blob/fe9a4f88713f7ee1692738ce0deb5935da362134/cpp/src/join/semi_join.cu#L92-L152

No, that implementation uses the old row comparator, which doesn't support nested types. If we want nested types, that implementation needs to be something like we just talked in the `cudf::contains` PR (https://github.com/rapidsai/cudf/pull/10656#discussion_r888484934).

> It would be incongruous to have the implementation for semi_join elsewhere when in it is so similar to the implementations for the other joins.

The problem is that,  *-joins firstly check for existence (using the same result generated from `cudf::contains`) then build a gather_map from that result. So we can use `cudf::contains` as a component to implement *-join.

If you want to keep the implementation inside *-join, then we will have `cudf::contains` internally call an external function like `detail::semi_anti_join_contains` which seems a bit awkward to me:
```
std::unique_ptr<column> contains(column_view const& haystack,
                                 column_view const& needles,
                                 rmm::cuda_stream_view stream,
                                 rmm::mr::device_memory_resource* mr)
{
  return cudf::detail::semi_anti_join_contains(haystack, needles, default_stream, mr);
}
```

**Comment by ttnghia** [comment]
So, we have 2 options:

1. Having *-joins call `cudf::contains`:
```
auto left_semi_anti_join(cudf::table_view const& left_keys, cudf::table_view const& right_keys,...) {

cudf::contains(left_keys, right_keys,....);
// build gather map
// return gather_map

```

2. Having `cudf::contains` calls `detail::semi_anti_join_contains`:
```
std::unique_ptr<column> contains(column_view const& haystack,
                                 column_view const& needles,
                                 rmm::cuda_stream_view stream,
                                 rmm::mr::device_memory_resource* mr)
{
  return cudf::detail::semi_anti_join_contains(haystack, needles, default_stream, mr);
}
```

**Comment by vyasr** [comment]
@ttnghia I see your point. The `semi_join` implementation is essentially 1) build a hash table on the right table, 2) probe the hash table with the left table entries and generate a boolean array indicating whether each row in left matches anything in right, and 3) perform a stream compaction using the boolean array to generate a gather map, which is returned. 

I think there are two conflicting considerations here. 
1. Steps 1+2 from above are essentially `contains`. That suggests that `semi_join` should depend on `contains`.
2. `contains` is a single-column search, i.e. the needles are scalars (perhaps of nested types) and the haystack is a column, whereas `semi_join` needs to support multiple columns. That suggests that `contains` should depend on contains.

Given these two considerations I think the second option that you outlined makes more sense. The function that you call `detail::semi_anti_join_contains` would need to have the signature
```
std::unique_ptr<column> contains(table_view const& haystack,
                                 table_view const& needles,
                                 rmm::cuda_stream_view stream,
                                 rmm::mr::device_memory_resource* mr)
```

If repeated lookups are valuable to optimize, we could extend the `hash_join` object to support holding either a `static_multimap` or a `static_map` as a way to also incorporate the `semi_/anti_join` logic into the same framework as the other joins. That would allow us to expose the containment check as a public API of `hash_join` object. That's probably premature at this stage however; implementing the above `contains(table_view, table_view)` helper function should be sufficient for now.

**Comment by vyasr** [comment]
To me that sounds like the last piece that I suggested: we may eventually want to implement something like a `static_map` version of the current `hash_join` object (which uses `static_multimap` for the other types of joins) in order to expose the hash lookup component for querying just the `contains` check. Using it for a single column would be suitable for `contains` or the various set-like operations (IIRC those are also all single columns of lists) while using it for multiple columns would be for semi/anti joins. I don't remember enough of @devavret's idea for #10409 to know if that would be consistent with what he had in mind, though.
